src/dune_lang/syntax.ml

```diff
@@ -44,22 +44,79 @@ module Version = struct
+  (* The extension supported versions are declared using an explicit list of all
+     versions but stored as a map from major versions to maps from minor version
+     to dune_lang required versions *)
+  type t = Version.t Int.Map.t Int.Map.t
+
+  let to_dyn t = Int.Map.to_dyn (Int.Map.to_dyn Version.to_dyn) t
+
+  (* We convert the exposed extension version type: `(Version.t * [ `Since of
+     Version.t ]) list` which is a list of fully qualified versions paired with
+     the corresponding dune_lang version. To the internal representation:
+     `(Version.t Int.Map.t) Int.Map.t` which is a list of major versions paired
+     with lists of minor versions paires with a dune_lang version. *)
+  let make (versions : (Version.t * [ `Since of Version.t ]) list) : t =
+    let v =
+      List.fold_left versions
+        ~init:(Int.Map.empty : Version.t Int.Map.t Int.Map.t)
+        ~f:(fun major_map (v_ext, `Since v_lang) ->
+          let major, minor = v_ext in
+          let minor_map =
+            match Int.Map.find major_map major with
+            | Some m -> m
+            | None -> Int.Map.empty
+          in
+          let minor_map = Int.Map.add_exn minor_map minor v_lang in
+          Int.Map.update major_map major ~f:(function
+            | Some _minor_map -> Some minor_map
+            | None -> Some minor_map))
+    in
+    v
+
+  let remove_uncompatible_versions lang_ver =
+    Int.Map.filter_map ~f:(fun minors ->
+        let minors =
+          Int.Map.filter minors ~f:(fun min_lang -> lang_ver >= min_lang)
+        in
+        if Int.Map.is_empty minors then
+          None
+        else
+          Some minors)
+
+  let rec greatest_supported_version ?lang_ver t =
+    let open Option.O in
+    match lang_ver with
+    | Some lang_ver ->
+      let compat = remove_uncompatible_versions lang_ver t in
+      greatest_supported_version compat
+    | None ->
+      let* major, minors = Int.Map.max_binding t in
+      let* minor, _ = Int.Map.max_binding minors in
+      Some (major, minor)
+
+  let get_min_lang_ver t (major, minor) =
+    let open Option.O in
+    let* minors = Int.Map.find t major in
+    Int.Map.find minors minor
+
+  let is_supported t (major, minor) lang_ver =
     match Int.Map.find t major with
-    | Some minor' -> minor' >= minor
+    | Some t -> (
+      match Int.Map.find t minor with
+      | Some min_lang_ver -> lang_ver >= min_lang_ver
+      | None -> false )
     | None -> false
 
-  let supported_ranges t =
-    Int.Map.to_list t
-    |> List.map ~f:(fun (major, minor) -> ((major, 0), (major, minor)))
+  let supported_ranges lang_ver (t : t) =
+    let compat = remove_uncompatible_versions lang_ver t in
+    Int.Map.to_list compat
+    |> List.map ~f:(fun (major, minors) ->
+           let max_minor, _ = Option.value_exn (Int.Map.max_binding minors) in
+           if major > 0 then
+             ((major, 0), (major, max_minor))
+           else
+             ((major, 1), (major, max_minor)))
```

Comment:
I don't understand why we do this.

Reply:
To prevent the printing of the range `0.0 to 0.4` while correctly printing `1.0 to 1.4`. But it was a quick hack, we could at least factorize the code better :-)

Reply:
why don't we want `0.0 to 0.4`?

Reply:
Because `0.0`is not a valid version number ? Or is it ?

Reply:
Yes, indeed. That makes sense.

Reply:
I changed a bit the code and added a comment.